buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java

```diff
@@ -233,41 +328,311 @@ protected ByteBuf newDirectBuffer(int initialCapacity, int maxCapacity) {
+     */
+    public static int defaultSmallCacheSize() {
+        return DEFAULT_SMALL_CACHE_SIZE;
+    }
+
+    /**
+     * Default normal cache size - System Property: io.netty.allocator.normalCacheSize - default 64
+     */
+    public static int defaultNormalCacheSize() {
+        return DEFAULT_NORMAL_CACHE_SIZE;
+    }
+
+    /**
+     * Return {@code true} if direct memory cache alignment is supported, {@code false} otherwise.
+     */
+    public static boolean isDirectMemoryCacheAlignmentSupported() {
+        return PlatformDependent.hasUnsafe();
+    }
+
     @Override
     public boolean isDirectBufferPooled() {
         return directArenas != null;
     }
 
-//    Too noisy at the moment.
-//
-//    public String toString() {
-//        StringBuilder buf = new StringBuilder();
-//        buf.append(heapArenas.length);
-//        buf.append(" heap arena(s):");
-//        buf.append(StringUtil.NEWLINE);
-//        for (PoolArena<byte[]> a: heapArenas) {
-//            buf.append(a);
-//        }
-//        buf.append(directArenas.length);
-//        buf.append(" direct arena(s):");
-//        buf.append(StringUtil.NEWLINE);
-//        for (PoolArena<ByteBuffer> a: directArenas) {
-//            buf.append(a);
-//        }
-//        return buf.toString();
-//    }
+    /**
+     * Returns {@code true} if the calling {@link Thread} has a {@link ThreadLocal} cache for the allocated
+     * buffers.
+     */
+    @Deprecated
+    public boolean hasThreadLocalCache() {
+        return threadCache.isSet();
+    }
+
+    /**
+     * Free all cached buffers for the calling {@link Thread}.
+     */
+    @Deprecated
+    public void freeThreadLocalCache() {
+        threadCache.remove();
+    }
+
+    final class PoolThreadLocalCache extends FastThreadLocal<PoolThreadCache> {
+        private final boolean useCacheForAllThreads;
+
+        PoolThreadLocalCache(boolean useCacheForAllThreads) {
+            this.useCacheForAllThreads = useCacheForAllThreads;
+        }
+
+        @Override
+        protected synchronized PoolThreadCache initialValue() {
+            final PoolArena<byte[]> heapArena = leastUsedArena(heapArenas);
+            final PoolArena<ByteBuffer> directArena = leastUsedArena(directArenas);
+
+            if (useCacheForAllThreads || Thread.currentThread() instanceof FastThreadLocalThread) {
+                return new PoolThreadCache(
+                        heapArena, directArena, tinyCacheSize, smallCacheSize, normalCacheSize,
+                        DEFAULT_MAX_CACHED_BUFFER_CAPACITY, DEFAULT_CACHE_TRIM_INTERVAL);
+            }
+            // No caching for non FastThreadLocalThreads.
+            return new PoolThreadCache(heapArena, directArena, 0, 0, 0, 0, 0);
+        }
+
+        @Override
+        protected void onRemoval(PoolThreadCache threadCache) {
+            threadCache.free();
+        }
+
+        private <T> PoolArena<T> leastUsedArena(PoolArena<T>[] arenas) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Consider using varargs for methods or constructors which take an array the last parameter. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)